README.md
# Early Construction ([mod portal][1])

![Screenshot](https://raw.githubusercontent.com/Aidiakapi/early_construction/master/screenshots/title.png)

Adds single use early game construction robots to greatly improve the quality of
life of base building. Uses the standard Factorio portable roboport mechanics.
Inspired by [Nanobots][2].

# Usage

1. Research the technology.
2. Craft the armor, equipment, and a bunch of robots.
3. Equip the armor.
4. Use blueprints.
5. ...
6. Profit?

# Issues?

If you run into any trouble, preferably submit an issue on GitHub, or otherwise
on the mod portal.

[1]: https://mods.factorio.com/mod/early_construction
[2]: https://mods.factorio.com/mod/Nanobots


src/changelog.md
# Changelog

## [0.9.3] - 2022-03-12

### Bugfixes

- Update Russian translation.

## [0.9.2] - 2022-02-28

### Features

- Add Russian translation by Eldrinn-Elantey.

## [0.9.1] - 2020-12-24

### Features

- Add base game's armor animations to the modded armors as well, thanks to a contribution by millerlarson.

## [0.9.0] - 2020-11-29

### Changes

- Update to Factorio 1.1.

## [0.8.1] - 2020-04-23

### Changes

- Re-modify graphics from base game, resulting in the same shadows as the base game, as well as updated visual effects.

### Bugfixes

- Fixed minor issue with the robot shadow. Thanks millerlarson for reporting!

## [0.8.0] - 2020-01-19

### Changes

- Update to Factorio 0.18.

## [0.7.0] - 2020-01-19

### Features

- Add compatibility with Industrial Revolution (different recipes).

## [0.6.0] - 2019-01-07

### Changes

- Early construction robots can now be deployed via non-personal roboports.
- Reimplemented the destruction logic to maximize mod compatibility, and simplified the code a lot.
- Early construction robots can now be mined by hand. It will return the robot to your inventory if it didn't place or mine anything.

## [0.5.5] - 2019-06-17

### Bugfixes

- Fixed crash introduced in last patch. Thanks schmorp for reporting!

## [0.5.4] - 2019-06-05

### Bugfixes

- Fixed that non-minable entity marked for deconstruction could result in an infinite stream of robots self-destructing. Thanks Nirahiel for reporting!

## [0.5.3] - 2019-06-01

### Bugfixes

- Fixed that you could sometimes lose some repair packs when a single robot both repaired and built/destroyed an entity. Thanks Nirahiel for reporting!

## [0.5.2] - 2019-05-13

### Bugfixes

- Fixed crash if the player died while the robots were out. Thanks Arcitos for reporting!

## [0.5.1] - 2019-05-10

### Bugfixes

- Upgrade planner will now return items. Thanks Nirahiel for reporting!

## [0.5.0] - 2019-05-08

### Features

- Destroyed entities will now leave behind a ghost after the first research (can be disabled in settings). Thanks Vandhal for the suggestion!

## [0.4.1] - 2019-05-04

### Bugfixes

- Fixed bug caused by mod API changes in 0.17.35. Thanks Vandhal for reporting!

## [0.4.0] - 2019-03-26

### Changes

- Robots are no longer killed, but instead destroyed. (They'll no longer show up under kills.)
- Destruction of robots is delayed by one frame to workaround a bug in Factorio that causes incompatibilities between mods. Thanks muppet9010 for reporting!

## [0.3.0] - 2019-02-23

### Changes

- Update to Factorio 0.17.

## [0.2.0] - 2018-11-19

### Features

- Add compatibility with Bob's Electronics (require basic circuit board for the first tier).

## [0.1.0] - 2018-11-19

### Features

- Add early game robots.


src/control.lua
local -- Forward declare functions
      reset_individual_global_per_player,
      on_player_created,
      on_player_respawned,
      on_player_removed,
      on_robot_built,
      on_robot_mined,
      on_built_entity,
      on_player_mined_entity,
      on_player_placed_equipment,
      on_player_removed_equipment,
      on_player_armor_inventory_changed,
      on_player_main_inventory_changed,
      spill_from_inventory_if_necessary,
      on_tick,
      has_disallowed_equipment,
      track_robot,
      untrack_robot,
      create_robot_death_effect,
      update_tick_handler,
      on_init,
      on_load,
      on_configuration_changed,
      on_configuration_changed_migrate_0_3_to_0_4,
      on_configuration_changed_migrate_0_5_to_0_6,
      on_configuration_changed_migrate_0_9_to_1_0,
      on_configuration_changed_handle_startup_setting_changes

reset_individual_global_per_player = function (player_index)
    local player = game.players[player_index]
    local character = player.character
    local grid = character and character.grid or nil
    global.per_player[player_index] = {
        last_armor_reclaimation_tick = 0,
        has_disallowed_equipment = grid and has_disallowed_equipment(grid),
    }
    spill_from_inventory_if_necessary(player_index)
end
on_player_created = function (event)
    reset_individual_global_per_player(event.player_index)
end
on_player_respawned = on_player_created
on_player_removed = function (event)
    global.per_player[event.player_index] = nil
end

on_robot_built = function (event)
    local robot = event.robot
    if robot.name ~= 'early-construction-robot' then return end
    track_robot(robot, event.tick + 1)
end

on_robot_mined = function (event)
    local robot = event.robot
    if robot.name ~= 'early-construction-robot' then return end
    track_robot(robot, event.tick + 1)
end

on_built_entity = function (event)
    local entity = event.created_entity
    if not entity or entity.name ~= 'early-construction-robot' then return end
    local player = game.players[event.player_index]

    entity.destroy()
    player.print({'early-construction-errors.placing'}, {r=1,g=0,b=0,a=1})
    player.insert({ name = 'early-construction-robot', count = 1 })
end

on_player_mined_entity = function (event)
    if not event.entity or not event.entity.valid or event.entity.name ~= 'early-construction-robot' then return end
    if global.tracked_robots[event.entity.unit_number] ~= nil then
        event.buffer.remove({ name = 'early-construction-robot', count = 1 })
        create_robot_death_effect(event.entity)
        untrack_robot(event.entity.unit_number)
    end
end

-- Prevents using early construction robots with regular personal roboports
has_disallowed_equipment = function (grid)
    for _, equipment in pairs(grid.equipment) do
        if equipment.type == 'roboport-equipment' and equipment.name ~= 'early-construction-equipment' then
            return true
        end
    end
    return false
end

local function on_player_modified_equipment_common(event)
    local grid = event.grid
    local player = game.players[event.player_index]

    local character = player.character
    local is_equipped = character ~= nil and character.grid == grid
    if not is_equipped then return end

    return player, character, grid
end

on_player_placed_equipment = function (event)
    local equipment = event.equipment
    if equipment.type ~= 'roboport-equipment' then return end
    if equipment.name == 'early-construction-equipment' then return end

    local player, character, grid = on_player_modified_equipment_common(event)
    if player == nil then return end

    local inventory = player.get_main_inventory()
    local robot_count = inventory.get_item_count('early-construction-robot')

    local per_player = global.per_player[player.index]
    if robot_count == 0 then
        per_player.has_disallowed_equipment = true
        return
    end

    player.print({'early-construction-errors.incompatible-equipment'}, {r=1,g=0,b=0,a=1})
    local item_stack = grid.take({ equipment = equipment, by_player = player })
    inventory.insert(item_stack)
end

on_player_removed_equipment = function (event)
    local player, character, grid = on_player_modified_equipment_common(event)
    if player == nil then return end

    local per_player = global.per_player[player.index]
    per_player.has_disallowed_equipment = has_disallowed_equipment(grid)
end

on_player_armor_inventory_changed = function (event)
    local player = game.players[event.player_index]

    local character = player.character
    if character == nil then return end

    local armor_slot = character.get_inventory(defines.inventory.character_armor)[1]
    local grid = armor_slot.valid and armor_slot.valid_for_read and armor_slot.grid or nil

    local per_player = global.per_player[player.index]
    if grid == nil or not has_disallowed_equipment(grid) then
        per_player.has_disallowed_equipment = false
        return
    end

    local inventory = player.get_main_inventory()
    local robot_count = inventory.get_item_count('early-construction-robot')
    if robot_count == 0 then
        per_player.has_disallowed_equipment = true
        return
    end
    
    player.print({'early-construction-errors.incompatible-equipment'}, {r=1,g=0,b=0,a=1})

    -- Put in inventory or drop on ground

    -- Clearing the player's hand location is necessary for properly swapping
    -- out armors. If you do not have this, then the game will "re-insert" the
    -- armor in the armor slot, causing an endless loop, and spilling them item
    -- instead.
    player.hand_location = nil

    -- Even with the previous line, there are still scenarios where there will
    -- be an "infinite loop" of the game picking up the item, and inserting it
    -- into the armor slot. If we already attempted to insert the armor into the
    -- inventory in the last 125ms, it falls back to spilling the armor into the
    -- world.
    local empty_stack = inventory.find_empty_stack(armor_slot.name)
    local per_player = global.per_player[player.index]
    if per_player.last_armor_reclaimation_tick + 8 < event.tick and
        empty_stack ~= nil and
        empty_stack.set_stack(armor_slot)
    then
        per_player.last_armor_reclaimation_tick = event.tick
        armor_slot.clear()
        return
    end

    character.surface.spill_item_stack(
        character.position, armor_slot,
        --[[enable_looted]] false, --[[force]] nil, --[[allow_belts]] false)
    armor_slot.clear()
end

on_player_main_inventory_changed = function (event)
    spill_from_inventory_if_necessary(event.player_index)
end

spill_from_inventory_if_necessary = function (player_index)
    local per_player = global.per_player[player_index]
    if not per_player.has_disallowed_equipment then return end

    local player = game.players[player_index]
    local character = player.character
    -- In theory, has_disallowed_equipment should always be false when there is
    -- no character, since a lack of character means a lack of equipment grid.
    -- In practice, there's probably some scenarios, such as right after a
    -- player dies, where this could occur, so better safeguard.
    if character == nil then return end

    local inventory = player.get_main_inventory()
    local robot_count = inventory.get_item_count('early-construction-robot')

    if robot_count == 0 then return end

    player.print({'early-construction-errors.robots-in-inventory'}, {r=1,g=0,b=0,a=1})
    local item_stack = { name = 'early-construction-robot', count = robot_count }
    character.surface.spill_item_stack(
        character.position, item_stack,
        --[[enable_looted]] false, --[[force]] nil, --[[allow_belts]] false)
    inventory.remove(item_stack)
end

on_tick = function (event)
    for unit_number, entry in pairs(global.tracked_robots) do
        if not entry.robot.valid then
            untrack_robot(unit_number)
        elseif event.tick >= entry.no_destruction_before_tick then
            local holds_items = false
            if entry.cargo_inventory.valid and not entry.cargo_inventory.is_empty() then holds_items = true end
            if entry.repair_inventory.valid and not entry.repair_inventory.is_empty() then holds_items = true end

            if not holds_items then
                untrack_robot(unit_number)
                create_robot_death_effect(entry.robot)
                entry.robot.destroy({
                    do_cliff_correction = false,
                    raise_destroy = true,
                })
            end
        end
    end
end

track_robot = function (robot, no_destruction_before_tick)
    if global.tracked_robots[robot.unit_number] then
        untrack_robot(robot.unit_number)
    end

    global.tracked_robots[robot.unit_number] = {
        robot = robot,
        no_destruction_before_tick = no_destruction_before_tick,
        cargo_inventory = robot.get_inventory(defines.inventory.robot_cargo),
        repair_inventory = robot.get_inventory(defines.inventory.robot_repair),
    }
    update_tick_handler()
end

untrack_robot = function (unit_number)
    if global.tracked_robots[unit_number] == nil then return end
    global.tracked_robots[unit_number] = nil
    update_tick_handler()
end

create_robot_death_effect = function (robot)
    robot.surface.create_entity({
        name = 'explosion-hit',
        position = robot.position,
        force = robot.force,
    })
end

update_tick_handler = function ()
    local should_be_ticking = table_size(global.tracked_robots) > 0
    global.is_ticking = should_be_ticking
    script.on_event(defines.events.on_tick, should_be_ticking and on_tick or nil)
end

on_init = function ()
    global.tracked_robots = {}
    global.tracked_robots_count = 0
    global.robots_pending_destruction = {}
    global.forces = {}
    for _, force in pairs(game.forces) do
        global.forces[force.name] = {}
    end
end

on_load = function ()
    if global.is_ticking then
        script.on_event(defines.events.on_tick, on_tick)
    end
end

on_configuration_changed = function (changes)
    if global.version == nil then
        on_configuration_changed_migrate_0_3_to_0_4()
        on_configuration_changed_migrate_0_5_to_0_6()
    end
    if global.version == 1 then
        on_configuration_changed_migrate_0_9_to_1_0()
    end

    if changes.mod_startup_settings_changed or changes.mod_changes['early_construction'] then
        on_configuration_changed_handle_startup_setting_changes()
    end
end

on_configuration_changed_migrate_0_3_to_0_4 = function ()
    -- Renamed global.tracked_robot_count to global.tracked_robots_count
    if global.tracked_robot_count then
        global.tracked_robot_count = nil
    end

    -- Previously, this variable wasn't initialized upon init, and was instead nil-checked
    -- now it is initialized. This variable can be safely derived from global.tracked_robots.
    if global.tracked_robots_count == nil then
        local count = 0
        for _ in pairs(global.tracked_robots) do
            count = count + 1
        end
        global.tracked_robots_count = count
    end

    -- Renamed global.ticking to global.is_ticking
    -- update_tick_handler invoked later will take care of initializing the global variable,
    -- and setting up the event handler if necessary.
    global.ticking = nil

    -- Newly introduced variables
    if global.robots_pending_destruction == nil then
        global.robots_pending_destruction = {}
    end

    update_tick_handler()
end

on_configuration_changed_migrate_0_5_to_0_6 = function ()
    local old_tracked_robots = global.tracked_robots
    for k, _ in pairs(global) do
        global[k] = nil
    end

    global.version = 1
    global.tracked_robots = {}
    for _, entry in pairs(old_tracked_robots) do
        local robot = entry.robot
        if (robot and
            robot.valid and
            robot.name == 'early-construction-robot' and
            global.tracked_robots[robot.unit_number] == nil) then

            global.tracked_robots[robot.unit_number] = {
                robot = robot,
                cargo_inventory = robot.get_inventory(defines.inventory.robot_cargo),
                repair_inventory = robot.get_inventory(defines.inventory.robot_repair),
                no_destruction_before_tick = game.tick + 1,
            }
        end
    end

    update_tick_handler()
end

on_configuration_changed_migrate_0_9_to_1_0 = function ()
    global.version = 2
    global.per_player = {}
    for _, player in pairs(game.players) do
        reset_individual_global_per_player(player.index)
    end
end

on_configuration_changed_handle_startup_setting_changes = function ()
    for _, force in pairs(game.forces) do
        if force.technologies['early-construction-light-armor'].researched then
            log(('[early_construction] resetting technology effects for force %q'):format(force.name))
            force.reset_technology_effects()
        end
    end
end

script.on_event(defines.events.on_player_created, on_player_created)
script.on_event(defines.events.on_player_respawned, on_player_respawned)
script.on_event(defines.events.on_player_removed, on_player_removed)
script.on_event(defines.events.on_robot_built_entity, on_robot_built)
script.on_event(defines.events.on_robot_built_tile, on_robot_built)
script.on_event(defines.events.on_robot_mined, on_robot_mined)
script.on_event(defines.events.on_built_entity, on_built_entity)
script.on_event(defines.events.on_player_mined_entity, on_player_mined_entity)
script.on_event(defines.events.on_player_placed_equipment, on_player_placed_equipment)
script.on_event(defines.events.on_player_removed_equipment, on_player_removed_equipment)
script.on_event(defines.events.on_player_armor_inventory_changed, on_player_armor_inventory_changed)
script.on_event(defines.events.on_player_main_inventory_changed, on_player_main_inventory_changed)

script.on_init(on_init)
script.on_load(on_load)
script.on_configuration_changed(on_configuration_changed)

-- commands.add_command('early_construction_dump_global', nil, function (command)
--     game.print('Global table for early_construction:')
--     game.print(serpent.block(global))
-- end)


src/data.lua
require("util")

local function combine_effects(...)
    local n = select("#", ...)
    local effects = {}
    for i = 1, n do
        local current = select(i, ...)
        if type(current) == "table" then
            for _, v in pairs(current) do
                effects[#effects + 1] = v
            end
        end
    end
    return effects
end

local ghosts_when_destroyed_effects
if settings.startup["early-construction-enable-entity-ghosts-when-destroyed"].value then
    ghosts_when_destroyed_effects = { {
        type = "ghost-time-to-live",
        modifier = 60 * 60 * 60 * 24 * 7
    } }
end

local light_armor_ingredients = {
    {"light-armor", 1},
    {"iron-plate", 10},
    {"iron-gear-wheel", 5},
    {"electronic-circuit", 40}
}
local heavy_armor_ingredients = {
    {"early-construction-light-armor", 1},
    {"heavy-armor", 1},
    {"electronic-circuit", 200},
    {"steel-plate", 20}
}
local equipment_ingredients = {
    {"electronic-circuit", 10}
}
local robot_ingredients = {
    {"repair-pack", 1 },
    {"coal", 2 }
}

local function patch_strings(from, to, table)
    for k, v in pairs(table) do
        if v == from then
            table[k] = to
        elseif type(v) == "table" then
            patch_strings(from, to, v)
        end
    end
end
--[[
    Bob's Electronics compatibility patch
]]
if data.raw.item["basic-circuit-board"] ~= nil then
    patch_strings("electronic-circuit", "basic-circuit-board", light_armor_ingredients)
    patch_strings("electronic-circuit", "basic-circuit-board", equipment_ingredients)
end

--[[
    Industrial Revolution compatibility patch
]]
if data.raw.technology["deadlock-bronze-age"] ~= nil then
    light_armor_ingredients = {
        {"light-armor", 1},
        {"copper-plate", 10},
        {"tin-gear-wheel", 5},
        {"copper-motor", 10}
    }
    heavy_armor_ingredients = {
        {"early-construction-light-armor", 1},
        {"heavy-armor", 1},
        {"copper-motor", 50},
        {"bronze-chassis-small", 1}
    }
    equipment_ingredients = {
        {"copper-motor", 5}
    }
    robot_ingredients = {
        {"tin-plate", 1},
        {"copper-gear-wheel", 1},
        {"coal", 2}
    }
end

local base_robot = data.raw['construction-robot']['construction-robot']
local function robot_clone_and_modify(value)
    local t = type(value)
    if t == 'table' then
        local new_value = {}
        for k, v in pairs(value) do
            new_value[robot_clone_and_modify(k)] = robot_clone_and_modify(v)
        end
        return new_value
    elseif t == 'string' then
        if value:find('__base__/graphics/entity/construction-robot/', 1, true) == 1
            and value:find('/', 45, true) == nil then
            return '__early_construction__/graphics/early-construction-robot/' .. value:sub(45)
        else
            return value
        end
    else
        return value
    end
end
local function robot_property(name)
    return robot_clone_and_modify(base_robot[name])
end

-- Apply armor type to character animation, borrowed from Simply Power Armor MK3
-- Contributed by millerlarson
for _, animation in ipairs(data.raw['character']['character']['animations']) do
    if animation.armors then
        for _, armor in ipairs(animation.armors) do
            if armor == 'light-armor' then
                animation.armors[#animation.armors + 1] = 'early-construction-light-armor'
            elseif armor == 'heavy-armor' then
                animation.armors[#animation.armors + 1] = 'early-construction-heavy-armor'
            end
        end
    end
end

data:extend(
    {
        -- Equipment
        {
            type = "equipment-category",
            name = "early-construction-armor"
        },
        {
            type = "item",
            name = "early-construction-equipment",
            icon = "__early_construction__/graphics/early-construction-equipment.png",
            icon_size = 32,
            placed_as_equipment_result = "early-construction-equipment",
            flags = {},
            subgroup = "equipment",
            order = "e[robotics]-a[early-construction-equipment]",
            stack_size = 5
        },
        {
            type = "roboport-equipment",
            name = "early-construction-equipment",
            take_result = "early-construction-equipment",
            sprite = {
                filename = "__early_construction__/graphics/early-construction-equipment.png",
                width = 32,
                height = 32,
                priority = "medium"
            },
            shape = {
                width = 1,
                height = 1,
                type = "full"
            },
            energy_source = {
                type = "electric",
                buffer_capacity = "0MJ",
                input_flow_limit = "0kW",
                usage_priority = "secondary-input"
            },
            charging_energy = "0kW",
            robot_limit = 15,
            construction_radius = 12,
            spawn_and_station_height = 0.4,
            charge_approach_distance = 2.6,
            recharging_animation = {
                filename = "__base__/graphics/entity/roboport/roboport-recharging.png",
                priority = "high",
                width = 37,
                height = 35,
                frame_count = 16,
                scale = 0.75,
                animation_speed = 0.5
            },
            recharging_light = {intensity = 0.4, size = 5},
            stationing_offset = {0, -0.6},
            charging_station_shift = {0, 0.5},
            charging_station_count = 0,
            charging_distance = 1.6,
            charging_threshold_distance = 5,
            categories = {"early-construction-armor"}
        },
        -- Armor
        {
            type = "equipment-grid",
            name = "small-early-construction-equipment-grid",
            width = 1,
            height = 1,
            equipment_categories = {"early-construction-armor"}
        },
        {
            type = "equipment-grid",
            name = "medium-early-construction-equipment-grid",
            width = 2,
            height = 1,
            equipment_categories = {"early-construction-armor"}
        },
        {
            type = "armor",
            name = "early-construction-light-armor",
            icon = "__early_construction__/graphics/light-armor.png",
            icon_size = 32,
            flags = {},
            resistances = {
                {
                    type = "physical",
                    decrease = 3,
                    percent = 20
                },
                {
                    type = "acid",
                    decrease = 0,
                    percent = 20
                },
                {
                    type = "explosion",
                    decrease = 2,
                    percent = 20
                },
                {
                    type = "fire",
                    decrease = 0,
                    percent = 10
                }
            },
            subgroup = "armor",
            equipment_grid = "small-early-construction-equipment-grid",
            order = "a[light-armor][early-construction]",
            stack_size = 1,
            infinite = true
        },
        {
            type = "armor",
            name = "early-construction-heavy-armor",
            icon = "__early_construction__/graphics/heavy-armor.png",
            icon_size = 32,
            flags = {},
            resistances = {
                {
                    type = "physical",
                    decrease = 6,
                    percent = 30
                },
                {
                    type = "explosion",
                    decrease = 20,
                    percent = 30
                },
                {
                    type = "acid",
                    decrease = 0,
                    percent = 40
                },
                {
                    type = "fire",
                    decrease = 0,
                    percent = 30
                }
            },
            subgroup = "armor",
            equipment_grid = "medium-early-construction-equipment-grid",
            order = "b[heavy-armor][early-construction]",
            stack_size = 1,
            infinite = true
        },
        -- Robot
        {
            type = "item",
            name = "early-construction-robot",
            icon = "__early_construction__/graphics/early-construction-robot.png",
            icon_size = 64, icon_mipmaps = 4,
            flags = {},
            subgroup = "logistic-network",
            order = "a[robot]-b[early-construction-robot]",
            place_result = "early-construction-robot",
            stack_size = 200
        },
        {
            type = "construction-robot",
            name = "early-construction-robot",
            icon = "__early_construction__/graphics/early-construction-robot.png",
            icon_size = 64, icon_mipmaps = 4,
            flags = {"placeable-player", "player-creation", "placeable-off-grid", "not-on-map"},
            minable = {hardness = 0.1, mining_time = 0.1, result = "early-construction-robot"},
            resistances = {{type = "fire", percent = 85}},
            max_health = 50,
            collision_box = {{0, 0}, {0, 0}},
            selection_box = {{-0.5, -1.5}, {0.5, -0.5}},
            hit_visualization_box = {{-0.1, -1.1}, {0.1, -1.0}},
            damaged_trigger_effect = robot_property('damaged_trigger_effect'),
            max_payload_size = 5,
            speed = 0.06,
            max_energy = "1MJ",
            energy_per_tick = "0kJ",
            speed_multiplier_when_out_of_energy = 1,
            energy_per_move = "0kJ",
            min_to_charge = 0.1,
            max_to_charge = 0.2,
            working_light = {intensity = 0.8, size = 3, color = {r = 0.85, g = 0.75, b = 0.75}},
            dying_explosion = "explosion",
            sparks = robot_property('sparks'),
            working_sound = robot_property('working_sound'),
            cargo_centered = {0.0, 0.2},
            construction_vector = {0.30, 0.22},
            water_reflection = robot_property('water_reflection'),
            idle = robot_property('idle'),
            idle_with_cargo = robot_property('idle_with_cargo'),
            in_motion = robot_property('in_motion'),
            in_motion_with_cargo = robot_property('in_motion_with_cargo'),
            shadow_idle = robot_property('shadow_idle'),
            shadow_idle_with_cargo = robot_property('shadow_idle_with_cargo'),
            shadow_in_motion = robot_property('shadow_in_motion'),
            shadow_in_motion_with_cargo = robot_property('shadow_in_motion_with_cargo'),
            working = robot_property('working'),
            shadow_working = robot_property('shadow_working'),
        },
        -- Recipes
        {
            type = "recipe",
            name = "early-construction-light-armor",
            enabled = false,
            energy_required = 3,
            ingredients = light_armor_ingredients,
            result = "early-construction-light-armor"
        },
        {
            type = "recipe",
            name = "early-construction-heavy-armor",
            enabled = false,
            energy_required = 8,
            ingredients = heavy_armor_ingredients,
            result = "early-construction-heavy-armor"
        },
        {
            type = "recipe",
            enabled = false,
            name = "early-construction-equipment",
            energy_required = 1,
            ingredients = equipment_ingredients,
            result = "early-construction-equipment"
        },
        {
            type = "recipe",
            name = "early-construction-robot",
            enabled = false,
            energy_required = 3,
            ingredients = robot_ingredients,
            result = "early-construction-robot",
            result_count = settings.startup["early-construction-robots-per-craft"].value
        },
        -- Technologies
        {
            type = "technology",
            name = "early-construction-light-armor",
            icon_size = 128,
            icon = "__early_construction__/graphics/technology.png",
            effects = combine_effects({
                {
                    type = "unlock-recipe",
                    recipe = "early-construction-robot"
                },
                {
                    type = "unlock-recipe",
                    recipe = "early-construction-light-armor"
                },
                {
                    type = "unlock-recipe",
                    recipe = "early-construction-equipment"
                },
            }, ghosts_when_destroyed_effects),
            unit = {
                count = 25,
                ingredients = {{"automation-science-pack", 1}},
                time = 5
            },
            order = "a-c-a"
        },
        {
            type = "technology",
            name = "early-construction-heavy-armor",
            icon_size = 128,
            icon = "__early_construction__/graphics/technology.png",
            effects = {
                {
                    type = "unlock-recipe",
                    recipe = "early-construction-heavy-armor"
                }
            },
            prerequisites = {"heavy-armor", "early-construction-light-armor", "logistic-science-pack"},
            unit = {
                count = 200,
                ingredients = {
                    {"automation-science-pack", 1},
                    {"logistic-science-pack", 1}
                },
                time = 30
            },
            order = "a-c-b"
        }
    }
)
